data_pipeline_api/file_api.py

```diff
@@ -15,12 +15,64 @@
 logger = getLogger(__name__)
 
 
-class AccessType:
-    """File access type constants.
+@dataclass(frozen=True)
+class FileAccess:
+    """Represents a generic file access.
     """
 
-    READ = "read"
-    WRITE = "write"
+    timestamp: datetime
+    call_metadata: Metadata
+    access_metadata: Metadata
+    path: Path
+
+    def to_access_log_record(
+        self, hash_cache: Optional[Dict[Path, str]] = None
+    ) -> Dict[str, Any]:
+        if hash_cache is None:
+            hash_cache = {}
+        access_metadata = self.access_metadata.copy()
+        if MetadataKey.calculated_hash not in access_metadata:
+            absolute_path = self.path.resolve()
+            try:
+                access_metadata[MetadataKey.calculated_hash] = hash_cache[absolute_path]
+            except KeyError:
+                access_metadata[MetadataKey.calculated_hash] = hash_cache[
+                    absolute_path
+                ] = FileAPI.calculate_hash(absolute_path)
```

Comment:
this is essentially doing like a lazy setdefault?

Reply:
yeah my aim is to avoid the computation unless absolutely necessary, because obviously for large files the hash calculation will be expensive. open to a better solution.